### DISCUSSION TITLE
[Mailer] How to change domain dynamically based on sender with mailgun (instead of static domain within MAILER_DSN) ?

### DISCUSSION THREAD

**Discussion by faizanakram99** [symfony/symfony#42369] [Q&A] [answered]
Our app is a multi-tenant app and as such has number of verified domains and we would like to send emails using selected tenant's domain.

A single tenant can have multiple verified domains and the actual domain which I would like to replace MAILER_DSN with would be known only at the time on composing emails (from address of email).

With swift mailer and cspoo swift mailer bundle, it was possible to configure domain dynamically, it doesn't seem possible with Symfony mailer.

**Comment by daum** [comment, marked as answer]
@faizanakram99 For this type of setup we generate the transport (ie `Transport::fromDsn($dsn)`) at run time based on the email being sent.  On the project where we do that, we have a custom message sent through the bus which then looks up based on the message and it's parameters which transport should be used, we pull that information from the database and create a new transport and mailer just for that message and send it via it.  

It doesn't sound like you do, but if you had a list of static transports you could predefine them like defined here: https://symfony.com/doc/current/mailer.html#multiple-email-transports and use the built in functionality.

**Reply by faizanakram99** [reply]
Thanks, yeah that's what I am doing now. Creating a factory which accepts "domain" and returns instance of `TransportInterface` (using `Transport::fromDsn`).

I was looking for a built-in way to do this, but apparently "domain" is private to mailgun transport and cannot be modified once service is instantiated.

**Reply by daum** [reply]
Yeah I didn't see a way around doing it that way.  You wouldn't want to be changing a transports properties after creation as once it is connected if you were to change for example it's connection details it wouldn't close/reopen the connection so you'd still be operating on the old one.  I think that is why majority of that stuff is all private/non-updatable.

**Reply by faizanakram99** [reply]
This is the point where mailgun endpoint is created and the method is invoked from `AbstractTransport::send` method (via some other sub-classes like `AbstractHttpTransport` and `AbstractApiTransport`), so had there been a way to modify domain dynamically, it probably would have been fine to call `send` after that.

https://github.com/symfony/mailgun-mailer/blob/5.3/Transport/MailgunApiTransport.php#L62
